google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java

```diff
@@ -371,14 +401,47 @@ void processReceivedMessages(List<ReceivedMessage> messages) {
         // we want to eventually
         // totally expire so that pubsub service sends us the message again.
         continue;
+      } else {
+        outstandingBatch.add(outstandingMessage);
       }
-      outstandingBatch.add(new OutstandingMessage(message, ackHandler));
       pendingReceipts.add(ackRequestData);
     }
-
     processBatch(outstandingBatch);
   }
 
+  synchronized void notifyAckSuccess(AckRequestData ackRequestData) {
+
+    if (outstandingReceipts.containsKey(ackRequestData.getAckId())) {
+      ReceiptCompleteData receiptCompleteData = outstandingReceipts.get(ackRequestData.getAckId());
+      OutstandingMessage outstandingMessage = receiptCompleteData.getOutstandingMessage();
+      // Setting to true means that the receipt is complete
+      receiptCompleteData.notifyReceiptComplete();
+      List<OutstandingMessage> outstandingBatch = new ArrayList<>();
+
+      for (Iterator<Entry<String, ReceiptCompleteData>> it =
+              outstandingReceipts.entrySet().iterator();
+          it.hasNext(); ) {
+        Map.Entry<String, ReceiptCompleteData> receipt = it.next();
+        // If receipt is complete then add to outstandingBatch to process the batch
+        if (receipt.getValue().isReceiptComplete()) {
+          it.remove();
+          if (pendingMessages.putIfAbsent(
+                  receipt.getKey(), receipt.getValue().getOutstandingMessage().ackHandler)
+              == null) { // msg2, msg1, msg3
```

Comment:
Remove this comment? // msg2, msg2, msg3

Reply:
oops yes, was just making sure logic was good.